### PR TITLE
tinygo: -compiler command line argument (e.g. tinygo)

### PR DESCRIPTION
**Pull request by archie2x**
help: 
- ~~I haven't figured out how to propagate this to mkuimage.~~
- it may make more sense as a buildopt instead of environ in build.go(?)
  - but the BuildOpts isn't always available
- name of flag? -go-compiler(?)
- any feedback! 

This allows specifying the compiler explicitly in makebb and handles peculiarities of tinygo.  It works by resolving and caching the path to the go-compiler then queries tinygo for its hidden go-build-tags and propagates them to package lookup / install. This allows the pruning algorithm to correctly add / remove files when syncing dependent packages.

Usage:
```
% makebb -h 
Usage of makebb:
  -compiler string
    	override go compiler to use (e.g. "/path/to/tinygo")
[...]
```
Resulting busybox is significantly smaller.  I compiled a slightly more complicated project than e.g. u-root/cmds/core/{true,false} to test package resolution.

Standard go:
```sh
u-root % GOOS=linux makebb cmds/core/init
10:10:54 Build environment: GOARCH=arm64 GOOS=linux GOPATH=/Users/roger/go CGO_ENABLED=0 GO111MODULE= GOROOT=/opt/homebrew/Cellar/go/1.22.2/libexec PATH=/opt/homebrew/Cellar/go/1.22.2/libexec/bin
10:10:54 Compiler: go version go1.22.2 darwin/arm64
10:10:59 Successfully built "bb" (size 2556039 bytes -- 2.4 MiB). <============
```

tinygo:
```sh
u-root % GOOS=linux makebb -compiler tinygo cmds/core/init
10:10:06 Build environment: GOARCH=arm64 GOOS=linux GOPATH=/Users/roger/go CGO_ENABLED=0 GO111MODULE= GOROOT=/opt/homebrew/Cellar/go/1.22.2/libexec PATH=/opt/homebrew/Cellar/go/1.22.2/libexec/bin
10:10:06 Compiler: tinygo version 0.33.0-dev-6184a6cd darwin/arm64 (using go version go1.22.2 and LLVM version 18.1.2)
10:10:20 Successfully built "bb" (size 1408888 bytes -- 1.3 MiB).  <============
```

Needs:
- tinygo dev branch (for syscall fixes supplied by @leongross )
- this patch to u-root to compile (link) with tinygo
- (rebased) tinygo-org/tinygo#4273

```
diff --git a/cmds/core/init/fns_other.go b/cmds/core/init/fns_other.go
index 850ce067..f979aaea 100644
--- a/cmds/core/init/fns_other.go
+++ b/cmds/core/init/fns_other.go
@@ -2,8 +2,8 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build !tinygo && !(darwin || dragonfly || freebsd || linux || netbsd || openbsd || solaris)
-// +build !tinygo,!darwin,!dragonfly,!freebsd,!linux,!netbsd,!openbsd,!solaris
+//go:build !(darwin || dragonfly || freebsd || linux || netbsd || openbsd || solaris)
+// +build !darwin,!dragonfly,!freebsd,!linux,!netbsd,!openbsd,!solaris
 
 package main
 
diff --git a/cmds/core/init/fns_unix.go b/cmds/core/init/fns_unix.go
index 4bcac715..4db8d9d0 100644
--- a/cmds/core/init/fns_unix.go
+++ b/cmds/core/init/fns_unix.go
@@ -2,8 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build !tinygo && (darwin || dragonfly || freebsd || linux || netbsd || openbsd || solaris)
-// +build !tinygo
+//go:build (darwin || dragonfly || freebsd || linux || netbsd || openbsd || solaris)
 // +build darwin dragonfly freebsd linux netbsd openbsd solaris
 
 package main
diff --git a/cmds/core/init/init_plan9.go b/cmds/core/init/init_plan9.go
index 4fb93f94..6b88f159 100644
--- a/cmds/core/init/init_plan9.go
+++ b/cmds/core/init/init_plan9.go
@@ -2,8 +2,8 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build !tinygo && plan9
-// +build !tinygo,plan9
+//go:build plan9
+// +build plan9
 
 package main
 
diff --git a/cmds/core/init/init_tinygo.go b/cmds/core/init/init_tinygo.go
new file mode 100644
index 00000000..1d670023
--- /dev/null
+++ b/cmds/core/init/init_tinygo.go
@@ -0,0 +1,29 @@
+//go:build tinygo
+
+package main
+
+/*
+ld.lld: error: undefined symbol: syscall.runtime_BeforeExec
+>>> referenced by exec_unix.go:282 ([...]/go/1.22.2/libexec/src/syscall/exec_unix.go:282)
+>>>               /var/folders/7f/_6cnvk8j475gw2rk8q3hv8q00000gn/T/tinygo3894403787/main.lto.main.o:(runtime.run$1$gowrapper)
+
+ld.lld: error: undefined symbol: syscall.runtime_AfterExec
+>>> referenced by exec_unix.go:308 ([...]/go/1.22.2/libexec/src/syscall/exec_unix.go:308)
+>>>               /var/folders/7f/_6cnvk8j475gw2rk8q3hv8q00000gn/T/tinygo3894403787/main.lto.main.o:(runtime.run$1$gowrapper)
+failed to run tool: ld.lld
+failed to link /var/folders/7f/_6cnvk8j475gw2rk8q3hv8q00000gn/T/tinygo3894403787/main: exit status 1
+*/
+
+import (
+	_ "unsafe" // for go:linkname
+)
+
+//go:linkname runtime_BeforeExec syscall.runtime_BeforeExec
+func runtime_BeforeExec() {
+
+}
+
+//go:linkname runtime_AfterExec syscall.runtime_AfterExec
+func runtime_AfterExec() {
+
+}
```

### CODE REVIEW

File: src/pkg/golang/build.go
Comment (leongross) [inline src/pkg/golang/build.go:278]: What do you think of refactoring this into smaller chunks and make it a bit more generic, such that the it becomes a bit more readable?

```go
type CompilerType int
const (
	CompilerGo CompilerType = iota
	CompilerTinygo
	CompilerUnkown
)

func CompilerTypeFromString(name string) CompilerType {
	val, ok := map[string]CompilerType{
		"go":     CompilerGo,
		"tinygo": CompilerTinygo,
	}[name]
	if ok {
		return val
	}
	return CompilerUnkown
}
```

Further one could implement a parsing function over the Compiler Type so that the parsing logic is put there.

```go
func (c *Compiler) Parse(cType CompilerType) (err error){
  switch cType {
    case CompilerGo:
      // parse
    
    case CompilerTinygo:
      // parse
    
    case CompilerUnkown:
      // parse
    
    default:
      return fmt.Errorf("unkown")
  }
}
```

```go
compilerID := CompilerTypeFromString(compiler.Identifier)
if err := compiler.Parse(compilerID); err != nil {
    // handle error
}
```

File: src/pkg/golang/build.go
Comment (leongross) [inline src/pkg/golang/build.go:437]: Same thing here, this could be separated from the main logic of the build function to make it more readable.

**Review by leongross — COMMENTED** [review]
Nice work, I anticipate this will help a lot to enable u-root builds with `tinygo`

**Comment by leongross** [comment]
@archie2x if you need said changes 
> diff --git a/cmds/core/init/fns_other.go b/cmds/core/init/fns_other.go

in u-root, do you mind opening a pr? at some point it needs to be merged anyway, right?

File: src/pkg/golang/build.go
Comment (archie2x) [reply src/pkg/golang/build.go:278]: I would be happy to! Thanks for the feedback!

[I'm just beginning to learn go]

**Comment by leongross** [comment]
@archie2x in this project you have to sign your commits (see [here](https://github.com/u-root/gobusybox/pull/121/checks?check_run_id=28430144629)). Without that, the CI is not run and with that reviewing changes might take longer.

**Comment by leongross** [comment]
@rminnich @hugelgupf what do you think of this?

File: src/pkg/golang/build.go
Comment (hugelgupf) [reply src/pkg/golang/build.go:278]: +1 on this

File: src/pkg/golang/build.go
Comment (hugelgupf) [inline src/pkg/golang/build.go:309]: tags := make(map[string]struct{})
for _, tag := range c.BuildTags {
  tags[tag] = struct{}{}
}
for _, tag := range info["build_tags"].([]string) {
  tags[tag] = struct{}{}
}
c.BuildTags = maps.Keys(tags) // import "maps"

File: src/pkg/golang/build.go
Comment (hugelgupf) [inline src/pkg/golang/build.go:265]: here's what I'd do for errors.

```
var ErrNoGoVersion = fmt.Errorf("go-compiler version output unrecognized")

...

return fmt.Errorf("%w: %s", ErrNoGoVersion, string(v))
```

This allows using errors.Is with `ErrNoGoVersion` and keeping the error message the same for the two efmt cases

File: src/cmd/makebb/makebb.go
Comment (leongross) [inline src/cmd/makebb/makebb.go:53]: This looks weird, probably one of my suggestions fits better?

```suggestion
	l.Printf("Build environment: %s\n", env)
	l.Printf("Compiler: %s\n", env.Compiler.VersionOutput)
```

Or
```suggestion
	l.Printf("Build environment: %s Compiler: %s\n", env, env.Compiler.VersionOutput)
```

File: src/pkg/golang/compiler.go
Comment (leongross) [inline src/pkg/golang/compiler.go:73]:
```suggestion
// If go-compiler is specified, resolve the absolute-path from PATH,
```

File: src/pkg/golang/compiler.go
Comment (leongross) [inline src/pkg/golang/compiler.go:91]:
```suggestion

```

File: src/pkg/golang/compiler.go
Comment (leongross) [inline src/pkg/golang/compiler.go:117]:
```suggestion

```

File: src/pkg/golang/compiler.go
Comment (leongross) [inline src/pkg/golang/compiler.go:99]: There is no occurance in your code that uses `v` as raw `[]bytes` returned by `cmd.CombinedOutput`. Consider making `v` a string and get rid of all the `string(v)` if you never need `v` itself anyways.

File: src/pkg/golang/compiler.go
Comment (leongross) [inline src/pkg/golang/compiler.go:175]:
```suggestion

```

File: src/pkg/golang/compiler.go
Comment (leongross) [inline src/pkg/golang/compiler.go:193]:
```suggestion

```

File: src/pkg/golang/compiler.go
Comment (leongross) [inline src/pkg/golang/compiler.go:216]:
```suggestion

```

File: src/pkg/golang/compiler.go
Comment (leongross) [inline src/pkg/golang/compiler.go:218]: I assume this is postponed? I am not sure about the options tinygo may offer here either. I will do some research about this, but independent of this PR, it should not be blocked by this. I think we can squeeze out some size reduction here by setting the right [build options](https://github.com/tinygo-org/tinygo/blob/release/compileopts/options.go#L22).

Comment (archie2x) [reply src/pkg/golang/compiler.go:218]: I think the issues is more that tinygo already does most of these by default so the existing command line arguments don't map well.

**Comment by leongross** [comment]
@hugelgupf I think @archie2x addressed all the requested changes, should we get this in?

**Comment by leongross** [comment]
Another thing: To enable tinygo support in u-root we had do move from go assembly to CGO. I tested this branch and realized that gobusybox does not support Cgo (also mentioned in the docs, I know). I tried to hack on this for a while but could not get it running yet.
I tried to build `exp/esxiboot` and got the following error:

```
14:27:07 Preserving bb generated source directory at /tmp/bb-2772405532 due to error: collecting and putting dependencies in place failed: writing package github.com/u-root/u-root/
pkg/boot/multiboot/internal/trampoline failed: $HOME/u-root/pkg/boot/multiboot/internal/
trampoline/trampoline_linux_amd64_tinygo.go:28:8: could not import C (no metadata for C)
```

I tracked the error down into [`go/packages/packages.go:1219`](https://cs.opensource.google/go/x/tools/+/master:go/packages/packages.go;l=1219?q=FakeImportC&ss=go%2Fx%2Ftools), but I am not sure where in the busybox builder I can check for the "C" package and ignore it.

**Review by leongross — APPROVED** [review]

**Comment by rminnich** [comment]
Leon approved these, it never merged, but if it passes, and looks reasonable to me, I'll merge.

**Pull request merged**
